doc/03_research/03_planning/06_state_machine_design.md

```diff
@@ -0,0 +1,278 @@
+    * [Brake](#brake)
+  * [Lane change state machine](#lane-change-state-machine)
+    * [DECIDE_LANE_CHANGE](#decidelanechange)
+    * [CHANGE_LANE_LEFT](#changelaneleft)
+    * [CHANGE_LANE_RIGHT](#changelaneright)
+  * [Intersection state machine](#intersection-state-machine)
+    * [APPROACH_INTERSECTION](#approachintersection)
+    * [IN_INTERSECTION](#inintersection)
+    * [TURN_LEFT](#turnleft)
+    * [STRAIGHT](#straight)
+    * [TURN_RIGHT](#turnright)
+    * [LEAVE_INTERSECTION](#leaveintersection)
+  * [Stop sign/traffic light state machine](#stop-signtraffic-light-state-machine)
+    * [STOP_NEAR](#stopnear)
+    * [STOP_SLOW_DOWN](#stopslowdown)
+    * [STOP_WILL_STOP](#stopwillstop)
+    * [STOP_WAIT](#stopwait)
+    * [STOP_GO](#stopgo)
+  * [Implementation](#implementation)
+    * [Sources](#sources)
+<!-- TOC -->
+
+## Super state machine
+
+![img.png](../../00_assets/Super_SM.png)
+
+The super state machine functions a controller of the main functions of the agent.
+
+Those functions are
+
+* following the road and brake in front of obstacles if needed
+* drive across an intersection
+* change lane
+
+## Driving state machine
+
+![img.png](../../00_assets/Driving_SM.png)
+
+Transition:
+
+* From `Intersection state machine`
+* From `Lane change state machine`
+
+This state machine controls the speed of the ego-vehicle. It either tells the acting part of the ego vehicle to `ACCELERATE`, `BRAKE` or `KEEP` the velocity.
+
+This is assumably the main state of the state machine as the ego vehicle will most of the time follow the road.
+
+### KEEP
+
+Transition:
+
+* From `ACCEL`
+* From `BRAKE`
+
+Keep the current target speed, applied most of the time. From here changes to the `ACCEL` or `BRAKE` state are performed, if events require a change of `target_speed`.
+
+### ACCEL
+
+Transition:
+
+* From `KEEP` if `new target_speed` is greater than current `target_speed`.
+* From `BRAKE` if `new target_speed` is greater than current `target_speed`.
+
+Set a new target speed and change back to `KEEP` state afterwards.
+
+### Brake
+
+Transition:
+
+* From `KEEP` if `new target_speed` is smaller than current `target_speed` or an `obstacle` or the `leading_vehicle` is in braking distance.
+* From `BRAKE` if `new target_speed` is smaller than current `target_speed` or an `obstacle` or the `leading_vehicle` is in braking distance.
+
+Set a new target speed and change back to `KEEP` state afterwards.
+
+## Lane change state machine
+
+![img.png](../../00_assets/Lane_Change_SM.png)
+
+Transition:
+
+* From `driving state machine` by `lane_change_requested`
+
+This state machine completes the change of a lane. This is triggered from the super state machine and can have multiple triggers. Those include:
+
+* Join highway
+* Leave highway
+* RoadOption:
+  * CHANGELANELEFT
+  * CHANGELANERIGHT
+  * KEEPLANE
+* avoid obstacle(doors, static objects)
+* give way to emergency vehicle
+* overtake slow moving vehicle
+* leave a parking bay
```

Comment:
Diese Option finde ich in der Grafik nicht

Reply:
Ebenfalls nur Auslöser, die unter lane_change_request zusammengefasst sind.